gapis/api/vulkan/query_timestamps.go

```diff
@@ -0,0 +1,437 @@
+	for i := uint32(0); i < submitCount; i++ {
+		si := submitInfos[i]
+		waitSemPtr := memory.Nullptr
+		waitDstStagePtr := memory.Nullptr
+		if count := uint64(si.WaitSemaphoreCount()); count > 0 {
+			waitSemPtr = allocAndRead(si.PWaitSemaphores().
+				Slice(0, count, l).
+				MustRead(ctx, cmd, s, nil)).Ptr()
+			waitDstStagePtr = allocAndRead(si.PWaitDstStageMask().
+				Slice(0, count, l).
+				MustRead(ctx, cmd, s, nil)).Ptr()
+		}
+
+		signalSemPtr := memory.Nullptr
+		if count := uint64(si.SignalSemaphoreCount()); count > 0 {
+			signalSemPtr = allocAndRead(si.PSignalSemaphores().
+				Slice(0, count, l).
+				MustRead(ctx, cmd, s, nil)).Ptr()
+		}
+
+		cmdBuffers := si.PCommandBuffers().Slice(0, uint64(si.CommandBufferCount()), s.MemoryLayout).MustRead(ctx, cmd, s, nil)
+		cmdCount := si.CommandBufferCount()
+		newCmdCount := cmdCount*2 + 1
+
+		commandbuffer := t.generateQueryCommand(ctx,
+			cb,
+			out,
+			query)
+		newCmdBUffers := make([]VkCommandBuffer, newCmdCount)
+		newCmdBUffers[0] = commandbuffer
+		for j := uint32(0); j < cmdCount; j++ {
+			buf := cmdBuffers[j]
+			newCmdBUffers[j*2+1] = buf
+
+			query++
+			commandbuffer = t.generateQueryCommand(ctx,
+				cb,
+				out,
+				query)
+			newCmdBUffers[j*2+2] = commandbuffer
+
+			begin := &path.Command{
+				Indices: []uint64{uint64(id), uint64(i), uint64(j), 0},
+			}
+			c, ok := GetState(s).CommandBuffers().Lookup(buf)
+			if !ok {
+				fmt.Errorf("Invalid command buffer %v", buf)
+			}
+			n := c.CommandReferences().Len()
+			end := &path.Command{
+				Indices: []uint64{uint64(id), uint64(i), uint64(j), uint64(n - 1)},
+			}
+			t.timestamps = append(t.timestamps, replay.Timestamps{Begin: begin, End: end})
+		}
+
+		cmdBufferPtr := allocAndRead(newCmdBUffers).Ptr()
+		newSubmitInfos[i] = NewVkSubmitInfo(s.Arena,
+			VkStructureType_VK_STRUCTURE_TYPE_SUBMIT_INFO,
+			0,                            // pNext
+			si.WaitSemaphoreCount(),      // waitSemaphoreCount
+			NewVkSemaphoreᶜᵖ(waitSemPtr), // pWaitSemaphores
+			NewVkPipelineStageFlagsᶜᵖ(waitDstStagePtr), // pWaitDstStageMask
+			newCmdCount,                        // commandBufferCount
+			NewVkCommandBufferᶜᵖ(cmdBufferPtr), // pCommandBuffers
+			si.SignalSemaphoreCount(),          // signalSemaphoreCount
+			NewVkSemaphoreᶜᵖ(signalSemPtr),     // pSignalSemaphores
+		)
+	}
+
+	submitInfoPtr := allocAndRead(newSubmitInfos).Ptr()
+
+	newCmd := cb.VkQueueSubmit(
+		cmd.Queue(),
+		cmd.SubmitCount(),
+		submitInfoPtr,
+		cmd.Fence(),
+		VkResult_VK_SUCCESS,
+	)
+
+	for _, read := range reads {
+		newCmd.AddRead(read.Data())
+	}
+	out.MutateAndWrite(ctx, id, newCmd)
+}
+
+func (t *queryTimestamps) Transform(ctx context.Context, id api.CmdID, cmd api.Cmd, out transform.Writer) {
+	s := out.State()
+	cb := CommandBuilder{Thread: cmd.Thread(), Arena: s.Arena}
+
+	defer func() {
+		for _, d := range t.allocated {
+			d.Free()
+		}
+	}()
+
+	switch cmd := cmd.(type) {
+	case *VkGetPhysicalDeviceProperties:
```

Comment:
It is not guaranteed that an application will call VkGetPhysicalDeviceProperties. We need another way of deciding what we should do for the timestamp period. I think we add the PhysicalDeviceProperties as an extra on the VkEnumeratePhysicalDevices, which means it should be in the state-block already.

Reply:
fixed